File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/data/db/repository/SpecieRepository.java
package com.dani_chuks.andeladeveloper.starwars.data.db.repository;

import android.support.annotation.NonNull;

import com.dani_chuks.andeladeveloper.starwars.dagger.ISchedulerProvider;
import com.dani_chuks.andeladeveloper.starwars.data.AppConstants;
import com.dani_chuks.andeladeveloper.starwars.data.SharedPreferenceManager;
import com.dani_chuks.andeladeveloper.starwars.data.db.DbUtils;
import com.dani_chuks.andeladeveloper.starwars.data.db.local.AppDatabase;
import com.dani_chuks.andeladeveloper.starwars.data.db.DataSource;
import com.dani_chuks.andeladeveloper.starwars.data.db.remote.ApiService;
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Specie;

import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;

public class SpecieRepository implements DataSource<Specie>{

    @NonNull
    private final ApiService apiService;
    @NonNull
    private final AppDatabase appDatabase;
    @NonNull
    private final ISchedulerProvider schedulerProvider;
    @NonNull
    private final CompositeDisposable disposableManager = new CompositeDisposable();
    private final SharedPreferenceManager preferenceManager;

    @Inject
    public SpecieRepository(@NonNull final ApiService apiService,
                            @NonNull final AppDatabase appDatabase,
                            @NonNull final ISchedulerProvider schedulerProvider,
                            final SharedPreferenceManager preferenceManager) {
        this.apiService = apiService;
        this.appDatabase = appDatabase;
        this.schedulerProvider = schedulerProvider;
        this.preferenceManager = preferenceManager;
    }

    @Override
    public Flowable<List<Specie>> getAll() {
        fetchIfEmpty();
        return appDatabase.specieDao().getAll();
    }

    @Override
    public Flowable<List<Specie>> getItemsLimitedToSize(final int size) {
        fetchIfEmpty();
        return appDatabase.specieDao().getItemBySize(size);
    }

    @Override
    public Flowable<List<Specie>> getAllAlphabetically() {
        fetchIfEmpty();
        return appDatabase.specieDao().getAllAlphabetically();
    }

    @Override
    public Flowable<Specie> getItemByUrl(final String stringUrl) {
        int specieId = DbUtils.getLastPathFromUrl(stringUrl);
        disposableManager.add(
                apiService.getSpecieById(specieId)
                        .subscribeOn(schedulerProvider.getIoScheduler())
                        .subscribe(specie -> appDatabase.specieDao().insertSpecie(specie)));

        return appDatabase.specieDao().getSpecieByUrl(stringUrl);
    }

    @Override
    public void insertItem(final Specie specie) {
        disposableManager.add(Observable.timer(1, TimeUnit.NANOSECONDS)
                .subscribeOn(schedulerProvider.getIoScheduler())
                .subscribe(time -> appDatabase.specieDao().insertSpecie(specie)));
    }

    @Override
    public void insertItemList(final List<Specie> species) {
        disposableManager.add(Observable.timer(1, TimeUnit.NANOSECONDS)
                .subscribeOn(schedulerProvider.getIoScheduler())
                .subscribe(time -> appDatabase.specieDao().insertSpecieList(species)));
    }

    private void fetchIfEmpty() {
        if (preferenceManager.isDataTypeFetchedOnce(AppConstants.SPECIE_RESOURCE_NAME)) {
            final int firstPage = 1;
            disposableManager.add(
                    apiService.getSpecieList(firstPage)
                            .subscribeOn(schedulerProvider.getIoScheduler())
                            .subscribe(response ->
                            {
                                appDatabase.specieDao().insertSpecieList(response.getSpecie());
                                preferenceManager.setResourceNextPage(AppConstants.SPECIE_RESOURCE_NAME, (firstPage + 1));
                                preferenceManager.setDataTypeFetchedOnce(AppConstants.SPECIE_RESOURCE_NAME);
                            }));
        }
    }
}

File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/data/db/repository/SpecieRepository.kt
package com.dani_chuks.andeladeveloper.starwars.data.db.repository

import com.dani_chuks.andeladeveloper.starwars.dagger.Result
import com.dani_chuks.andeladeveloper.starwars.data.AppConstants
import com.dani_chuks.andeladeveloper.starwars.data.SharedPreferenceManager
import com.dani_chuks.andeladeveloper.starwars.data.db.DbUtils
import com.dani_chuks.andeladeveloper.starwars.data.db.local.AppDatabase
import com.dani_chuks.andeladeveloper.starwars.data.models.SpecieList
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Specie
import com.dani_chuks.andeladeveloper.starwars.home.HomeViewModel.Companion.firstPage
import io.reactivex.disposables.CompositeDisposable
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import javax.inject.Inject

class SpecieRepository @Inject
constructor(private val appDatabase: AppDatabase,
            private val remoteDataSource: SpecieRemoteDataSource,
            private val preferenceManager: SharedPreferenceManager) {
    private val disposableManager = CompositeDisposable()

    suspend fun all(): List<Specie>  {
            return appDatabase.specieDao().all()
        }

    suspend fun allAlphabetically(): List<Specie> = coroutineScope {
        appDatabase.specieDao().allAlphabetically()
    }

    suspend fun getItemsLimitedToSize(size: Int): List<Specie> = coroutineScope {
        appDatabase.specieDao().getItemBySize(size)
    }

    suspend fun getItemByUrl(stringUrl: String): Specie = coroutineScope {
        val specie = async { appDatabase.specieDao().getSpecieByUrl(stringUrl) }.await()
        if(specie == null){
            val specieId = DbUtils.getLastPathFromUrl(stringUrl)
            val specieFromRemote = remoteDataSource.getPersonById(specieId)
            if(specieFromRemote is Result.Success) {
                specieFromRemote.data.let { appDatabase.specieDao().insert(it) }
            }
        }
        specie
    }

    suspend fun insertItem(t: Specie) {
        appDatabase.specieDao().insert(t)
    }

    suspend fun insertItemList(t: List<Specie>) = coroutineScope{
        appDatabase.specieDao().insertList(t)
    }

    suspend fun fetchAndSync(page: Int): Result<SpecieList> = coroutineScope {
        val speciesFetched = remoteDataSource.getSpeciesFromPage(page)
        println("fetch and sync Species from page $page ==> $speciesFetched")
        if (speciesFetched is Result.Success) {
            speciesFetched.data.specie?.let {
                appDatabase.specieDao().insertList(it)
                preferenceManager.setResourceNextPage(AppConstants.SPECIE_RESOURCE_NAME, firstPage + 1)
                preferenceManager.setDataTypeFetchedOnce(AppConstants.SPECIE_RESOURCE_NAME)
            }
        }
        speciesFetched
    }

}
